### PR TITLE
Selfie: Import count down view from Camera2

### LINKED ISSUE #93: Replace timer chip with CountDownView
See: https://github.com/LineageOS/android_packages_apps_Camera2/blob/lineage-19.1/src/com/android/camera/ui/CountDownView.java

### PR DESCRIPTION
**Pull request by luca020400**
Converted to kotlin with some changes here and
there

Signed-off-by: Luca Stefani <luca.stefani.ge1@gmail.com>
Change-Id: I068daa661de3ea478c719f890506202d2bb2df84

### CODE REVIEW

File: app/src/main/java/org/lineageos/selfie/ui/CountDownView.kt [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    }
+    private var mRemainingSecs = 0
+    private lateinit var mListener: () -> Unit
+    private val mHandler = MainHandler(Looper.getMainLooper())
+    private val mPreviewArea = Rect()
+
+    /**
+     * Returns whether countdown is on-going.
+     */
+    private val isCountingDown = mRemainingSecs > 0
```
Comment (SebaUbuntu) [inline app/src/main/java/org/lineageos/selfie/ui/CountDownView.kt]: I think this should be inside a get() function, unless Kotlin treats that assignment as a unit but I doubt that

Comment (luca020400) [reply app/src/main/java/org/lineageos/selfie/ui/CountDownView.kt]: We aren't using this bit, but yeah, fixed.

File: app/src/main/res/layout/activity_main.xml [hunk 2]
```diff
@@ -180,18 +180,22 @@
...
-        app:rippleColor="#00ffffff" />
+        app:layout_constraintTop_toTopOf="parent">
+
+        <TextView
+            android:id="@+id/remaining_seconds"
+            android:layout_width="match_parent"
+            android:layout_height="match_parent"
+            android:gravity="center"
+            android:textColor="@android:color/white"
+            android:textSize="160sp" />
```
Comment (luca020400) [inline app/src/main/res/layout/activity_main.xml]: This needs to be decreased.

File: app/src/main/java/org/lineageos/selfie/MainActivity.kt [hunk 3]
```diff
@@ -104,7 +105,7 @@ class MainActivity : AppCompatActivity() {
     private val settingsButton by lazy { findViewById<ImageButton>(R.id.settingsButton) }
     private val shutterButton by lazy { findViewById<ImageButton>(R.id.shutterButton) }
     private val timerButton by lazy { findViewById<ImageButton>(R.id.timerButton) }
-    private val timerChip by lazy { findViewById<Chip>(R.id.timerChip) }
+    private val countDownView by lazy { findViewById<CountDownView>(R.id.count_down_view) }
```
Comment (luk1337) [inline app/src/main/java/org/lineageos/selfie/MainActivity.kt]: sort pls

File: app/src/main/java/org/lineageos/selfie/ui/CountDownView.kt [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+import androidx.annotation.IntRange
+import org.lineageos.selfie.R
+
+/**
+ * This class manages the looks of the countdown.
+ */
+class CountDownView(context: Context, attrs: AttributeSet?) : FrameLayout(
+    context, attrs
+) {
+    private val mRemainingSecondsView by lazy {
```
Comment (luk1337) [inline app/src/main/java/org/lineageos/selfie/ui/CountDownView.kt]: nuke m prefix

File: app/src/main/java/org/lineageos/selfie/ui/CountDownView.kt [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+     */
+    fun onPreviewAreaChanged(previewArea: Rect) {
+        mPreviewArea.set(previewArea)
+    }
+
+    private fun remainingSecondsChanged(seconds: Int) {
+        mRemainingSecs = seconds
+        if (seconds == 0) {
+            // Countdown has finished.
+            visibility = INVISIBLE
```
Comment (luk1337) [inline app/src/main/java/org/lineageos/selfie/ui/CountDownView.kt]: isInvisible = true

File: app/src/main/java/org/lineageos/selfie/ui/CountDownView.kt [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        val textHeight = mRemainingSecondsView.measuredHeight
+        mRemainingSecondsView.scaleX = 1f
+        mRemainingSecondsView.scaleY = 1f
+        mRemainingSecondsView.translationX = mPreviewArea.centerX() - textWidth / 2f
+        mRemainingSecondsView.translationY = mPreviewArea.centerY() - textHeight / 2f
+        mRemainingSecondsView.pivotX = textWidth / 2f
+        mRemainingSecondsView.pivotY = textHeight / 2f
+        mRemainingSecondsView.alpha = 1f
+        val endScale = 2.5f
+        mRemainingSecondsView.animate().scaleX(endScale).scaleY(endScale)
```
Comment (luk1337) [inline app/src/main/java/org/lineageos/selfie/ui/CountDownView.kt]: would probably add some newlines here

File: app/src/main/java/org/lineageos/selfie/ui/CountDownView.kt [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+        val textWidth = mRemainingSecondsView.measuredWidth
+        val textHeight = mRemainingSecondsView.measuredHeight
+        mRemainingSecondsView.scaleX = 1f
+        mRemainingSecondsView.scaleY = 1f
+        mRemainingSecondsView.translationX = mPreviewArea.centerX() - textWidth / 2f
+        mRemainingSecondsView.translationY = mPreviewArea.centerY() - textHeight / 2f
+        mRemainingSecondsView.pivotX = textWidth / 2f
+        mRemainingSecondsView.pivotY = textHeight / 2f
+        mRemainingSecondsView.alpha = 1f
+        val endScale = 2.5f
```
Comment (luk1337) [inline app/src/main/java/org/lineageos/selfie/ui/CountDownView.kt:76]: const

File: app/src/main/java/org/lineageos/selfie/ui/CountDownView.kt [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    /**
+     * Starts showing countdown in the UI.
+     *
+     * @param sec duration of the countdown, in seconds
+     */
+    fun startCountDown(@IntRange(from = 0) sec: Int) {
+        if (isCountingDown) {
+            cancelCountDown()
+        }
+        visibility = VISIBLE
```
Comment (luk1337) [inline app/src/main/java/org/lineageos/selfie/ui/CountDownView.kt]: isVisible = true

File: app/src/main/java/org/lineageos/selfie/ui/CountDownView.kt [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    /**
+     * Cancels the on-going countdown in the UI, if any.
+     */
+    private fun cancelCountDown() {
+        if (mRemainingSecs > 0) {
+            mRemainingSecs = 0
+            mHandler.removeMessages(SET_TIMER_TEXT)
+            visibility = INVISIBLE
```
Comment (luk1337) [inline app/src/main/java/org/lineageos/selfie/ui/CountDownView.kt]: isInvisible = true

File: app/src/main/java/org/lineageos/selfie/ui/CountDownView.kt [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        }
+    }
+
+    private inner class MainHandler(looper: Looper) : Handler(looper) {
+        override fun handleMessage(message: Message) {
+            if (message.what == SET_TIMER_TEXT) {
```
Comment (luk1337) [inline app/src/main/java/org/lineageos/selfie/ui/CountDownView.kt]: would use when() even tho we handle only one case

File: app/src/main/java/org/lineageos/selfie/ui/CountDownView.kt [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+                remainingSecondsChanged(mRemainingSecs - 1)
+            }
+        }
+    }
+
+    companion object {
+        private const val SET_TIMER_TEXT = 1
+        private const val ANIMATION_DURATION_MS = 800L
+    }
+}
```
Comment (luk1337) [inline app/src/main/java/org/lineageos/selfie/ui/CountDownView.kt]: add newline

File: app/src/main/java/org/lineageos/selfie/ui/CountDownView.kt [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+        mRemainingSecondsView.alpha = 1f
+        val endScale = 2.5f
+        mRemainingSecondsView.animate().scaleX(endScale).scaleY(endScale)
+            .alpha(0f).setDuration(ANIMATION_DURATION_MS).start()
+    }
+
+    /**
+     * Sets a listener that gets notified when the status of countdown has finished.
+     */
+    fun setCountDownStatusListener(listener: () -> Unit) {
```
Comment (luk1337) [inline app/src/main/java/org/lineageos/selfie/ui/CountDownView.kt:88]: just make listener public?

File: app/src/main/res/layout/activity_main.xml [hunk 13]
```diff
@@ -180,18 +180,22 @@
...
-        android:id="@+id/timerChip"
-        android:layout_width="wrap_content"
-        android:layout_height="wrap_content"
-        android:layout_marginTop="8dp"
-        android:layout_marginEnd="16dp"
-        android:textColor="@color/white"
-        android:visibility="gone"
-        app:chipSurfaceColor="@color/black"
+    <org.lineageos.selfie.ui.CountDownView
+        android:id="@+id/count_down_view"
```
Comment (luk1337) [inline app/src/main/res/layout/activity_main.xml]: camelCase

File: app/src/main/res/layout/activity_main.xml [hunk 14]
```diff
@@ -180,18 +180,22 @@
...
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        android:visibility="invisible"
         app:layout_constraintEnd_toEndOf="parent"
-        app:layout_constraintTop_toTopOf="@+id/viewFinder"
-        app:rippleColor="#00ffffff" />
+        app:layout_constraintTop_toTopOf="parent">
+
+        <TextView
+            android:id="@+id/remaining_seconds"
```
Comment (luk1337) [inline app/src/main/res/layout/activity_main.xml]: camelCase

**Review by luk1337 — APPROVED** [review]

**Review by SebaUbuntu — APPROVED** [review]

**Pull request merged**